// protojs/lib/dotjs/methods/utils.js
'use strict';
var slice = Array.prototype.slice;
/**
 * - [times](#times)
 * - [repeat](#repeat)
 * - [tap](#tap)
 * - [result](#result)
 * - [identity](#identity)
 * - [property](#property)
 * - [compareProperty](#compareProperty)
 * - [noop](#noop)
 */
var utilMethods = module.exports = {
    times: times,
    repeat: repeat,
    tap: tap,
    result: result,
    identity: identity,
    property: property,
    compareProperty: compareProperty,
    noop: noop
};
/**
 * Calls `callback` `self` times with `thisArg` as context. Callback is passed iteration index from 0 to `self-1`
 * 
 * @param {Integer} self
 * @param {Function} callback
 * @param {Any} thisArg
 * @return {Array}
 */
function times(callback, thisArg) {
    var self = this.self;
    var arr = Array(Math.max(0, self));
    for (var i = 0; i < self; i++) arr[i] = callback.call(thisArg, i);
    this.self = arr;
    return this;
}
/**
 * Returns array with the first argument repeated `times` times
 * @param  {Any} self
 * @param  {Integer} times
 * @return {Array}
 */
function repeat(times) {
    var self = this.self;
    var arr = Array(Math.max(0, times));;
    for (var i = 0; i < times; i++) arr[i] = self;
    this.self = arr;
    return this;
}
/**
 * Function to tap into chained methods and to inspect intermediary result
 *
 * @param {Any} self value that's passed between chained methods
 * @param {Function} func function that will be called with the value (both as context and as the first parameter)
 * @return {Any}
 */
function tap(func) {
    var self = this.self;
    func.call(self, self);
    this.self = self;
    return this;
};
/**
 * Calls function `self` (first parameter of _.result) with given context and arguments
 * 
 * @param {Function|Any} self
 * @param {Any} thisArg context
 * @param {List} arguments extra arguments
 * @return {Any}
 */
function result(thisArg) { //, arguments
    var args = slice.call(arguments, 1);
    var result = typeof self == 'function' ? self.apply(thisArg, args) : self;
    this.self = result;
    return this;
}
/**
 * Returns self. Useful for using as an iterator if the actual value needs to be returned. Unlike in underscore and lodash, this function is NOT used as default iterator.
 *
 * @param {Any} self 
 * @return {Any}
 */
function identity() {
    var self = this.self;
    this.self = self;
    return this;
}
/**
 * Returns function that picks the property from the object
 *
 * @param {String} self
 * @return {Function}
 */
function property() {
    var self = this.self;
    var func = function(obj) {
        return obj[self];
    };
    this.self = func;
    return this;
}
/**
 * Returns function that can be used in array sort to sort by a given property
 *
 * @param {String} self
 * @return {Function}
 */
function compareProperty() {
    var self = this.self;
    var func = function(a, b) {
        return a[self] < b[self] ? -1 : a[self] > b[self] ? 1 : 0;
    };
    this.self = func;
    return this;
}
/**
 * Function that does nothing
 */
function noop() {
    this.self = undefined;
    return this;
}
